### PR TITLE
HIVE-24535: Cleanup AcidUtils.Directory and remove unnecessary filesystem listing

### PR DESCRIPTION
**Pull request by pvargacl**
### What changes were proposed in this pull request?

- AcidUtils.getAcidState is doing a recursive listing on S3 FileSystem, it already knows the content of each delta and base directory, this could be returned to OrcInputFormat, to avoid listing each delta directory again there.
- AcidUtils.getAcidstate submethods are collecting more and more infos about the state of the data directory. This could be done directly to the final Directory object to avoid 10+ parameters in methods.
- AcidUtils.Directory, OrcInputFormat.AcidDirInfo and AcidUtils.TxnBase can be merged to one class, to clean up duplications.

### Why are the changes needed?
 - performance improvement

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing unit tests

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 1]
```diff
@@ -994,18 +857,76 @@ public long getVisibilityTxnId() {
...
+          + writeId + "; visibilityTxnId: " + visibilityTxnId;
+    }
+  }
+  /**
+   * In addition to {@link ParsedBaseLight} this knows if the data is in raw format, i.e. doesn't
+   * have acid metadata columns embedded in the files.  To determine this in some cases
+   * requires looking at the footer of the data file which can be expensive so if this info is
+   * not needed {@link ParsedBaseLight} should be used.
+   */
+  public static final class ParsedBase extends ParsedBaseLight {
```
Comment (klcopp) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:888]: Seems like this might kind of duplicate AcidBaseFileInfo? If so we can get rid of AcidBaseFileInfo?

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 2]
```diff
@@ -1253,19 +1190,12 @@ public static ParsedDelta parsedDelta(Path deltaDir, String deltaPrefix, FileSys
...
             .collect(Collectors.toList());
-      } else if (isDeleteDelta) {
-        // For delete deltas we need the files for AcidState
-        try {
-          files = SHIMS.listLocatedHdfsStatus(fs, deltaDir, bucketFileFilter);
-        } catch (UnsupportedOperationException uoe) {
-          files = Arrays.stream(fs.listStatus(deltaDir, bucketFileFilter))
-              .map(HdfsFileStatusWithoutId::new)
-              .collect(Collectors.toList());
-        }
```
Comment (klcopp) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1268]: Why was this block (1260-1268) needed originally?

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 3]
```diff
@@ -1490,49 +1432,31 @@ else if (prev != null && next.maxWriteId == prev.maxWriteId
...
-    }
-
-    Path base = null;
-    boolean isBaseInRawFormat = false;
-    if (bestBase.basePath != null) {
-      base = bestBase.basePath;
-      isBaseInRawFormat = MetaDataFile.isRawFormat(base, fs, dirSnapshots != null ? dirSnapshots.get(base) : null);
+       * For now, assert that ValidTxnList.VALID_TXNS_KEY is set only if this is really a read
+       * of a transactional table.
+       * see {@link #getChildState(FileStatus, HdfsFileStatusWithId, ValidWriteIdList, List, List, List, List, TxnBase, boolean, List, Map, FileSystem, ValidTxnList)}
```
Comment (klcopp) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1454]: getChildState parameters have been changed

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 4]
```diff
@@ -3360,14 +3257,13 @@ public static Directory getAcidStateFromCache(Supplier<FileSystem> fileSystem,
...
     if (recompute || (value == null)) {
-      Directory dirInfo = getAcidState(fileSystem.get(), candidateDirectory, conf,
+      AcidDirectory dirInfo = getAcidState(fileSystem.get(), candidateDirectory, conf,
           writeIdList, useFileIds, ignoreEmptyFiles);
       value = new DirInfoValue(writeIdList.writeToString(), dirInfo);
 
       if (value.dirInfo != null && value.dirInfo.getBaseDirectory() != null
           && value.dirInfo.getCurrentDirectories().isEmpty()) {
         if (dirCacheDuration > 0) {
-          populateBaseFiles(dirInfo, useFileIds, fileSystem);
```
Comment (klcopp) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3370]: Why was this here?

**Review by klcopp — COMMENTED** [review]
Stupid question – where does this skip a filesystem call? Aren't the FS calls in OrcInputFormat.FileGenerator#callInternalmoved just moved to AcidDirectory#getBaseAndDeltaFiles?

**Comment by pvargacl** [comment]
The FileSystem calls are saved because on S3 AcidUtils#getAcidState will do a recursive listing (getHdfsDirSnapshots), and when it processes a delta directory, it will store the file list in ParsedDelta, so when AcidDirectory#getBaseAndDeltaFiles is called later the files are returned without a second FS call. If we are on HDFS there is no gain, the getBaseAndDeltaFiles will still do a listing as it did before in OrcInputFormat.FileGenerator#callInternalmoved.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidDirectory.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+   */
+  public List<AcidUtils.ParsedDelta> getDeleteDeltas() {
+    return currentDirectories.stream().filter(AcidUtils.ParsedDeltaLight::isDeleteDelta).collect(Collectors.toList());
+  }
+
+  /**
+   * All original, base and delta bucket files that should be read by this reader
+   * @throws IOException ex
+   */
+  public List<AcidUtils.AcidBaseFileInfo> getBaseAndDeltaFiles() throws IOException {
```
[the remark this answers is not in the document]
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidDirectory.java:188]: No, this is the main file list that is used for split generation. It is one of the input parameter of OrcInputFormat#determineSplitStrategies.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidDirectory.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    List<AcidUtils.AcidBaseFileInfo> baseAndDeltaFiles = new ArrayList<>();
+    if (base == null) {
```
[the remark this answers is not in the document]
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidDirectory.java:190]: If it is a normal acid table and contains only a few delta, this branch will do nothing, because there can not be original files in the table (the delta files will be added later from the currentDirectories). This path is here (if I understood it correctly) for tables that were converted to transactional, but not yet compacted.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java (hunk 1 above)
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:888]: ParsedBase represents a base directory with many files. AcidBaseFileInfo name is rather confusing for me, but it represent any datafile that could be in an acid table (original, bucketfile in base, bucketfile in delta) These are the "base" files for orc splits.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java (hunk 2 above)
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1268]: For delete delta folders the metadata for all files will be send to the execution side, so the FileId could be used to retrieve the orctail - and with the new feature of Adam - the whole file from llap cache. This was here to ensure the file list is propagated even in HDFS where we don't have anything in directory snapshot. Now this functionality is hidden inside ParsedDelta#getFiles it will either return the list from cache or do the FileSystem call itself if it was not done before

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java (hunk 4 above)
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:3370]: The basefiles were populated and stored in cache, so they could be later used in OrcInputFormat if no write happens to the table before the next query. This is not necessary now, because ParsedBase will handle this when getFiles is called, if the baseFiles are not there it will do the listing once and store it. What is even better, if we are on S3 the files will be there already from the first recursive listing, so the FS is not even called once.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidDirectory.java [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+      }
+    }
+
+    return baseAndDeltaFiles;
+  }
+
+  @Override
+  public String toString() {
+    return "Aborted Directories: " + abortedDirectories + "; original: " + originalFiles + "; obsolete: " + obsolete
+        + "; currentDirectories: " + currentDirectories + "; base: " + base;
```
[the remark this answers is not in the document]
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidDirectory.java:227]: Will do that, I will add the rawformat to ParsedBase#toString, this will print that too.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java (hunk 3 above)
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1454]: Removed this link, because it is outdated, i don't see anything related to this in getChildState.

**Comment by klcopp** [comment]
Got it, thanks!

**Review by klcopp — APPROVED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcInputFormat.java [hunk 8]
```diff
@@ -1316,7 +1282,7 @@ private Directory getAcidState() throws IOException {
     }
 
 
-    private AcidDirInfo callInternal() throws IOException {
+    private AcidDirectory callInternal() throws IOException {
       if (context.acidOperationalProperties != null
```
[the remark this answers is not in the document]
Comment (pvargacl) [reply ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcInputFormat.java:1286]: I played around this a bit and I think will do a follow up ticket for mm, because this again is very messy.
For mm tables HiveInputformat will take the partition/table directory, find the delta directories in it and validate them against the validwriteid list. This is neccessary to handle every other file format. But the consequence of this, that the OrcInputFormat#getSplits will be called with the delta directories directly, not with the partition/table dir as it is done normally. So this bypass actually will do a listing in the delta dir itself. And "findOriginals" not only finds the original for MM tables it finds all of the files in the table, and it is necessary otherwise we would not generate split for normal bucketfiles. So the sort answer is that these originals is not the same originals as the context of mm tables ...
The most disgusting part is, when you create an orc table that is not transactional, this method will be called, and we will create a full AcidDirectory just to list all the files in the table as originals. OrcInputformat is very much interweaved with acid code.

[the remark this answers is not in the document]
Comment (klcopp) [reply ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcInputFormat.java:1286]: Thanks for taking the time to research & explain this!
